src/vs/workbench/contrib/terminal/browser/terminalService.ts

```diff
@@ -645,7 +655,25 @@ export class TerminalService implements ITerminalService {
 			return;
 		}
 		const tabs = this._terminalGroupService.groups.map(g => g.getLayoutInfo(g === this._terminalGroupService.activeGroup));
-		const state: ITerminalsLayoutInfoById = { tabs };
+
+		// Save terminals in editors too
+		const seenPersistentProcessIds: number[] = [];
+		for (const t of tabs) {
+			for (const term of t.terminals) {
+				seenPersistentProcessIds.push(term.terminal);
+			}
+		}
+		const otherInstances = this.instances.filter(instance => typeof instance.persistentProcessId === 'number' && instance.shouldPersist && seenPersistentProcessIds.indexOf(instance.persistentProcessId) === -1);
```

Comment:
It would be simpler to get the terminals in the editor area from `terminalEditorService`

Reply:
tried and that doesn't work, but should. 

I believe it doesn't work because of this, which happens when you close the window and then the state is saved with no terminals
https://github.com/ssigwart/vscode/blob/a084f63f4d3e5477efbcd8851008e8d996a45681/src/vs/workbench/contrib/terminal/browser/terminalEditorService.ts#L92-L103